Add python pandoc-xnos

ppenguin opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

`pandoc-xnos` is a python-based `pandoc` filter (consisting of `pandoc-xnos` which can be called by separate binaries for cross-references to equations, tables, sections, figures) when converting from markdown with pandoc.

See upstream: https://github.com/tomduck/pandoc-xnos

The binary installed can be called with `pandoc --filter pandoc-xnos`.

To use only a specific filter (subset of `xnos`) install the related packages `pandoc-{eq|fig|table|sec}nos` also included in this PR.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/executable-nix-shell-for-directly-calling-a-binary-from-pip/18181/5

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/pandoc-omatic-and-filters-install-location-issues/18179/2

fabaff commented on pkgs/development/python-modules/pandoc-eqnos/default.nix:
```diff
@@ -0,0 +1,27 @@
+  };
+
+  propagatedBuildInputs = [ pandoc-xnos ];
+
```
The upstream repo contains [tests](https://github.com/tomduck/pandoc-eqnos/tree/master/test) but it looks like that they need to be generated on the fly.

```suggestion

  # Different pandoc executables are not available
  doCheck = false;

  pythonImportsCheck = [
    "pandoc_eqnos"
  ];
```

fabaff commented on pkgs/development/python-modules/pandoc-eqnos/default.nix:
```diff
@@ -0,0 +1,27 @@
+  propagatedBuildInputs = [ pandoc-xnos ];
+
+  meta = with lib; {
+    description = "Pandoc-eqnos is a pandoc filter for numbering equations and their references when converting from markdown to other formats.";
```
```suggestion
    description = "Pandoc filter for numbering equations and their references";
```

see https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes

fabaff commented on pkgs/development/python-modules/pandoc-eqnos/default.nix:
```diff
@@ -0,0 +1,27 @@
+  meta = with lib; {
+    description = "Pandoc-eqnos is a pandoc filter for numbering equations and their references when converting from markdown to other formats.";
+    homepage = "https://github.com/tomduck/pandoc-eqnos";
+    license = licenses.gpl3;
```
See https://github.com/tomduck/pandoc-eqnos/blob/master/setup.py#L6-L8

```suggestion
    license = licenses.gpl3Only;
```

fabaff reviewed (commented):
Could you please create a commit per module with `python3Packages.xxx: init at x.x.x`?

The comments are also valid for the other modules.

ppenguin committed:
add self to maintainers

ppenguin commented:
> Could you please create a commit per module with `python3Packages.xxx: init at x.x.x`?
> 
> The comments are also valid for the other modules.

done, thanks for the guidance.

loicreynier commented:
> @fabaff Oops, it was spotted in #164552 that these filters should probably be packaged as `pythonApplication` (not `pythonPackage`), since they are not intended to use via `import`. I'll make the necessary changes (hopefully this week, quite busy now) and change the respective commits and PR to that effect.

I agree that individual filters should be packaged as `pythonApplication` but `pandoc-xnos` is also a library and may be used as a package in some projects. In that regards, should not it be build as  a `pythonPackage` and as `pythonApplication`?

ppenguin commented:
> 



> > @fabaff Oops, it was spotted in #164552 that these filters should probably be packaged as `pythonApplication` (not `pythonPackage`), since they are not intended to use via `import`. I'll make the necessary changes (hopefully this week, quite busy now) and change the respective commits and PR to that effect.
> 
> I agree that individual filters should be packaged as `pythonApplication` but `pandoc-xnos` is also a library and may be used as a package in some projects. In that regards, should not it be build as a `pythonPackage`?

Also a good point... However the only ones that actually seem to use it as a library are the packages that are part of the `pandoc-xnos-suite` (i.e. `pandoc-eqnos` etc.). Since I suppose the main use case is using `pandoc-xnos` directly as a filter (only in cases where you'd specifically want to use *only* e.g. `eqnos` *and not* the others you'd use a single filter from the suite), the predominant use would probably still be as a standalone filter.

ppenguin commented:
Ok, that should be it: `pandoc-{eq,seq,fig,table}nos` moved to `tools/misc` as `pythonApplication`, `pandoc-xnos` remains as `python3Package`.

loicreynier commented on pkgs/development/python-modules/pandoc-xnos/default.nix:
```diff
@@ -0,0 +1,32 @@
+  pythonImportsCheck = [ "pandocxnos" ];
+
+  meta = with lib; {
+    description = "The pandoc-xnos filter suite provides facilities for cross-referencing in markdown documents processed by pandoc.";
```
```suggestion
    description = "Pandoc filter suite providing facilities for cross-referencing in markdown documents";
```

loicreynier commented on pkgs/tools/misc/pandoc-eqnos/default.nix:
```diff
@@ -0,0 +1,30 @@
+  doCheck = false;
+
+  meta = with lib; {
+    description = "Pandoc filter for numbering equations and their references";
```
```suggestion
    description = "Standalone Pandoc filter from the pandoc-xnos suite for numbering equations and equation references";
```

loicreynier commented on pkgs/tools/misc/pandoc-fignos/default.nix:
```diff
@@ -0,0 +1,30 @@
+  doCheck = false;
+
+  meta = with lib; {
+    description = "Pandoc filter for numbering figures and their references.";
```
```suggestion
    description = "Standalone Pandoc filter from the pandoc-xnos suite for numbering figures and figure references";
```

loicreynier commented on pkgs/tools/misc/pandoc-secnos/default.nix:
```diff
@@ -0,0 +1,33 @@
+  doCheck = false;
+
+  meta = with lib; {
+    description = "Pandoc filter for numbering sections and their references.";
```
```suggestion    
    description = "Standalone Pandoc filter from the pandoc-xnos suite for numbering section references";
```

loicreynier commented on pkgs/tools/misc/pandoc-tablenos/default.nix:
```diff
@@ -0,0 +1,30 @@
+  doCheck = false;
+
+  meta = with lib; {
+    description = "Pandoc filter for numbering tables and their references.";
```
```suggestion
    description = "Standalone Pandoc filter from the pandoc-xnos suite for numbering tables and table references";
```

loicreynier reviewed (commented):
I left some `meta.description` suggestions to fit the [guidelines](https://nixos.org/manual/nixpkgs/stable/#var-meta-description) and to emphasis that `pandoc-`(`eqnos`, `fignos`, `secnos` and `tablenos`) are standalone filters.

ppenguin commented:
@fabaff sorry for prodding, you all must be busy with 22.05 too... Is there anything remaining for me to do or is this one merge-ready now? (there indeed have been some more changes since last review, but all addressed issues and some others were addressed in the meantime...)

Cheers, ppenguin

SuperSandro2000 reviewed (commented):
Please squash the commits together to one per package and one for the maintainer addition.

ppenguin commented on pkgs/tools/misc/pandoc-tablenos/default.nix:
```diff
@@ -0,0 +1,30 @@
+
+  propagatedBuildInputs = [ pandoc-xnos ];
+
+  # Different pandoc executables are not available
+  doCheck = false;
```
If I understand correctly these are simply stubs calling `xnos`, so the reason to keep them separate upstream (pip) would merely be to select fine-grained capability, even though from a resource-usage perspective it hardly makes a difference.
But I guess it's cleanest to just mirror the package units/scope from upstream.

ppenguin committed:
python3Packages.pandoc-xnos: init at 2.5.0

ppenguin committed:
pandoc-eqnos: init at 2.5.0

ppenguin committed:
pandoc-fignos: init at 2.4.0

ppenguin committed:
pandoc-secnos: init at 2.2.2

ppenguin committed:
pandoc-tablenos: init at 2.3.0

ppenguin commented on pkgs/tools/misc/pandoc-secnos/default.nix:
```diff
@@ -0,0 +1,33 @@
+    url = "https://github.com/ppenguin/pandoc-secnos.git";
+    rev = "165ee1f4c1208636254392335d34934dc50d273e";
+    sha256 = "sha256-5VgEq5NeKRexPLJyQ8Uwj3vfnN9LUGoVIUUMiaWzyxw=";
+  };
```
done

SuperSandro2000 commented on pkgs/development/python-modules/pandoc-xnos/default.nix:
```diff
@@ -0,0 +1,32 @@
+
+  propagatedBuildInputs = [ pandocfilters psutil ];
+
+  doCheck = true;
+
```
```suggestion
```
We don't need to repeat the default.

ppenguin committed:
Update pkgs/development/python-modules/pandoc-xnos/default.nix

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>
